### PR TITLE
DROOLS-4680: [DMN Designer] BC DOs as DMN DTs - Create backend service that receives a Data Object and generate a Data Type

### PR DESCRIPTION
**Pull request by manstis**
See https://issues.jboss.org/browse/DROOLS-4680

As discussed in our Scrum meeting yesterday the UI and service only supports top level Data Objects at present. i.e. no nested types.

### CODE REVIEW

**Comment by danielzhe** [comment]
For me it's good. I already have the conversion to `DataObject` to `DataType` working in my local branch, so if we merge it asap I can "glue" this with the UI and we all can be happy. \o/

About the UI, should I keep the entire package name like the image bellow or should I show only the classname? I'm asking because when I convert it `DataObject` to `DataType` I'll use only the class name (not the full name with package) but it could lead to name clash (for example, if you have `com.MyType` and `com.mortgages.MyType`. What do you think? @manstis 

image

**Comment by manstis** [comment]
@danielzhe I think we should preserve the FQCN in `DataObject` and handle conflicts in the UI _if_ the User chooses to import both `com.MyType` and `com.mortgages.MyType`. Both would, as you say, become `MyType` within the namespace of the diagram in which they're imported; but there would still be a conflict that would need resolving. If there is such a conflict you could (1) simply append `1`, `2` etc to the resulting `ItemDefinition` name or (2) present the User with a dialog to provide new names etc.

**Comment by manstis** [comment]
Jenkins please retest this.

**Comment by jomarko** [comment]
```
22:07:47 [ERROR] Failed to execute goal org.apache.maven.plugins:maven-enforcer-plugin:3.0.0-M1:enforce (ban-duplicated-classes) on project kie-wb-common-dmn-webapp-kogito-common: Some Enforcer rules have failed. Look above for specific messages explaining why the rule failed. -> [Help 1]
```

**Comment by jomarko** [comment]
Manually I didn't reveal issues.

File: kie-wb-common-dmn/kie-wb-common-dmn-backend/src/main/java/org/kie/workbench/common/dmn/backend/editors/types/DataObjectsServiceImpl.java [hunk 1]
```diff
@@ -13,45 +13,157 @@
...
+            if (Objects.nonNull(builtInType)) {
+                return builtInType.getName();
+            }
+        } catch (ClassNotFoundException cnfe) {
+            //Swallow as BuiltInType.ANY is the default response
+        }
+        return BuiltInType.ANY.getName();
+    }
+
+    private String unbox(final String typeName) {
```
[the remark this answers is not in the document]
Comment (manstis) [reply kie-wb-common-dmn/kie-wb-common-dmn-backend/src/main/java/org/kie/workbench/common/dmn/backend/editors/types/DataObjectsServiceImpl.java]: Done. Moved to `DataModelOracleUtilities`

File: kie-wb-common-dmn/kie-wb-common-dmn-backend/src/main/java/org/kie/workbench/common/dmn/backend/editors/types/DataObjectsServiceImpl.java [hunk 2]
```diff
@@ -13,45 +13,157 @@
...
+        } else if (LocalTime.class.isAssignableFrom(clazz) || OffsetTime.class.isAssignableFrom(clazz)) {
+            return BuiltInType.TIME;
+        } else if (ZonedDateTime.class.isAssignableFrom(clazz) || OffsetDateTime.class.isAssignableFrom(clazz) || LocalDateTime.class.isAssignableFrom(clazz)) {
+            return BuiltInType.DATE_TIME;
+        } else if (Boolean.class.isAssignableFrom(clazz)) {
+            return BuiltInType.BOOLEAN;
+        } else if (Map.class.isAssignableFrom(clazz)) {
+            return BuiltInType.CONTEXT;
+        }
+        return BuiltInType.ANY;
```
[the remark this answers is not in the document]
Comment (manstis) [reply kie-wb-common-dmn/kie-wb-common-dmn-backend/src/main/java/org/kie/workbench/common/dmn/backend/editors/types/DataObjectsServiceImpl.java]: I cannot move this to `BuiltInType` as that is used client-side and server-side and some of the classes in this method are not available to GWT (client-side). This method therefore needs to live on the server-side only.

File: kie-wb-common-dmn/kie-wb-common-dmn-webapp-kogito-common/src/test/java/org/kie/workbench/common/dmn/webapp/kogito/common/backend/workarounds/MockDataObjectsServiceImplTest.java [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+    private MockDataObjectsServiceImpl dataObjectsService;
+
+    @Before
+    public void setup() {
+        this.dataObjectsService = new MockDataObjectsServiceImpl();
+    }
+
+    @Test
+    public void testLoadDataObjectsWithNullWorkspaceProject() {
+        assertTrue(dataObjectsService.loadDataObjects(null).isEmpty());
```
[the remark this answers is not in the document]
Comment (manstis) [reply kie-wb-common-dmn/kie-wb-common-dmn-webapp-kogito-common/src/test/java/org/kie/workbench/common/dmn/webapp/kogito/common/backend/workarounds/MockDataObjectsServiceImplTest.java]: Done.

File: kie-wb-common-dmn/kie-wb-common-dmn-webapp-kogito-common/src/test/java/org/kie/workbench/common/dmn/webapp/kogito/common/backend/workarounds/MockDataObjectsServiceImplTest.java [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+    }
+
+    @Test
+    public void testLoadDataObjectsWithNonNullWorkspaceProject() {
+        assertTrue(dataObjectsService.loadDataObjects(workspaceProject).isEmpty());
```
[the remark this answers is not in the document]
Comment (manstis) [reply kie-wb-common-dmn/kie-wb-common-dmn-webapp-kogito-common/src/test/java/org/kie/workbench/common/dmn/webapp/kogito/common/backend/workarounds/MockDataObjectsServiceImplTest.java]: Done.

File: kie-wb-common-dmn/kie-wb-common-dmn-webapp-standalone/src/test/java/org/kie/workbench/common/dmn/showcase/backend/workarounds/MockDataObjectsServiceImplTest.java [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+    private MockDataObjectsServiceImpl dataObjectsService;
+
+    @Before
+    public void setup() {
+        this.dataObjectsService = new MockDataObjectsServiceImpl();
+    }
+
+    @Test
+    public void testLoadDataObjectsWithNullWorkspaceProject() {
+        assertTrue(dataObjectsService.loadDataObjects(null).isEmpty());
```
[the remark this answers is not in the document]
Comment (manstis) [reply kie-wb-common-dmn/kie-wb-common-dmn-webapp-standalone/src/test/java/org/kie/workbench/common/dmn/showcase/backend/workarounds/MockDataObjectsServiceImplTest.java]: Done.

File: kie-wb-common-dmn/kie-wb-common-dmn-webapp-standalone/src/test/java/org/kie/workbench/common/dmn/showcase/backend/workarounds/MockDataObjectsServiceImplTest.java [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+    }
+
+    @Test
+    public void testLoadDataObjectsWithNonNullWorkspaceProject() {
+        assertTrue(dataObjectsService.loadDataObjects(workspaceProject).isEmpty());
```
[the remark this answers is not in the document]
Comment (manstis) [reply kie-wb-common-dmn/kie-wb-common-dmn-webapp-standalone/src/test/java/org/kie/workbench/common/dmn/showcase/backend/workarounds/MockDataObjectsServiceImplTest.java]: Done.

File: kie-wb-common-dmn/kie-wb-common-dmn-backend/src/test/java/org/kie/workbench/common/dmn/backend/editors/types/DataObjectsServiceImplTest.java [hunk 7]
```diff
@@ -16,70 +16,472 @@
...
+                                                      ModelField.FIELD_ORIGIN.SELF,
+                                                      FieldAccessorsAndMutators.BOTH,
+                                                      DataType.TYPE_NUMERIC_SHORT),
+                                       new ModelField("doubleField",
+                                                      Double.class.getName(),
+                                                      ModelField.FIELD_CLASS_TYPE.REGULAR_CLASS,
+                                                      ModelField.FIELD_ORIGIN.SELF,
+                                                      FieldAccessorsAndMutators.BOTH,
+                                                      DataType.TYPE_NUMERIC_DOUBLE)
+                               });
```
[the remark this answers is not in the document]
Comment (manstis) [reply kie-wb-common-dmn/kie-wb-common-dmn-backend/src/test/java/org/kie/workbench/common/dmn/backend/editors/types/DataObjectsServiceImplTest.java]: Done.

**Comment by jomarko** [comment]
@manstis once you push changes, please start full build.

**Comment by manstis** [comment]
Jenkins please execute full downstream build.

**Comment by manstis** [comment]
@jomarko Updates pushed and FDB triggered.

**Comment by manstis** [comment]
Looks like I have some work to do on this one...

**Comment by manstis** [comment]
Jenkins please execute full downstream build.

**Comment by manstis** [comment]
Code that I wrote vs simply formatted has at least 94% unit test coverage. I was not planning on writing unit tests for the classes I formatted (after moving the constants they use). I hope this is acceptable.

**Comment by danielzhe** [comment]
My PR is ready, I'm just waiting this one to be merged before open my PR.

**Comment by manstis** [comment]
Just need @jomarko approval but he's on PTO today.

**Pull request merged**